src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp

```diff
@@ -559,47 +637,97 @@ void ParquetBlockInputFormat::initializeIfNeeded()
+                    nested_column = nullable_column->getNestedColumnPtr();
+                }
+
+                return parquetTryHashColumn(nested_column.get(), parquet_column_descriptor);
+            };
+
+            key_condition_with_bloom_filter_data->prepareBloomFilterData(hash_one, hash_many);
+
+            filtering_columns = getBloomFilterFilteringColumnKeys(key_condition_with_bloom_filter_data->getRPN());
+        }
     }
 
-    for (int row_group = 0; row_group < num_row_groups; ++row_group)
+    auto skip_row_group_based_on_filters = [&](int row_group)
     {
-        if (skip_row_groups.contains(row_group))
-            continue;
+        if (!format_settings.parquet.filter_push_down && !format_settings.parquet.bloom_filter_push_down)
+        {
+            return false;
+        }
+
+        KeyCondition::ColumnIndexToBloomFilter column_index_to_bloom_filter;
+
+        const auto & header = getPort().getHeader();
 
-        if (parquet_bloom_filter_condition)
+        std::vector<Range> hyperrectangle(header.columns(), Range::createWholeUniverse());
+
+        if (format_settings.parquet.filter_push_down)
         {
-            const auto column_index_to_bf = buildColumnIndexToBF(*bf_reader, row_group, index_mapping, filtering_columns);
+            hyperrectangle = getHyperrectangleForRowGroup(*metadata, row_group, header, format_settings);
+        }
 
-            if (!parquet_bloom_filter_condition->mayBeTrueOnRowGroup(column_index_to_bf))
-            {
-                continue;
-            }
+        if (format_settings.parquet.bloom_filter_push_down)
+        {
+            column_index_to_bloom_filter = buildColumnIndexToBF(*bf_reader, row_group, index_mapping, filtering_columns);
         }
 
-        if (format_settings.parquet.filter_push_down && key_condition
-            && !key_condition
-                    ->checkInHyperrectangle(
-                        getHyperrectangleForRowGroup(*metadata, row_group, getPort().getHeader(), format_settings),
-                        getPort().getHeader().getDataTypes())
-                    .can_be_true)
+        bool maybe_exists = key_condition_with_bloom_filter_data->checkInHyperrectangle(hyperrectangle, getPort().getHeader().getDataTypes(), column_index_to_bloom_filter).can_be_true;
+
+        return !maybe_exists;
+    };
+
+    for (int row_group = 0; row_group < num_row_groups; ++row_group)
+    {
+        if (skip_row_groups.contains(row_group))
             continue;
 
-        if (row_group_batches.empty() || row_group_batches.back().total_bytes_compressed >= min_bytes_for_seek)
+        if (key_condition_with_bloom_filter_data && skip_row_group_based_on_filters(row_group))
+        {
+            continue;
+        }
+
+        // When single-threaded parsing, can prefetch row groups, so need to put all row groups in the same row_group_batch
+	if (row_group_batches.empty() || row_group_batches.back().total_bytes_compressed >= min_bytes_for_seek)
```

Comment:
Why leading tab?

Reply:
Hm.. I suspect it is because I edited this line using `vi` and it somehow messed up with identation.